core/src/eval/macros.rs

```diff
@@ -55,6 +55,21 @@ macro_rules! push_u256 {
 	)
 }
 
+/// Pops top element of the stack and converts it to `usize`.
+///
+/// The top element **must** be not greater than `usize::MAX`.
+/// This non-local invariant is enforced by gas metering infrastructure.
+macro_rules! pop_usize {
```

Comment:
This is probably a larger refactor, and likely will have no impact so I doubt we'll do it, but out of curiosity, what is your opinion about having all these `pop_*` macros as opposed to functions?

As I understand it, in the places where we write

```rust
pop_usize!(state, var);
```

it should be equivalent to

```rust
let var = pop_usize(state)?;
```

for some appropriate definition of `pop_usize` (and perhaps some implementation of `std::ops::Try` on whatever the return type of `pop_usize` is; I'm imaging something like `Either<usize, Control>`).

To my eye, the function version looks nicer because the assignment is explicit in the syntax, whereas the macro version looks like a random side-effecting call. But maybe all the hoops you have to jump through to make the function version work with the short-circuiting is just not worth it?

Reply:
I agree, makes sense as a function rather than a macro. Macros are great if you need something that is bit more general blanketing of Structs. Or, if you need a function that needs to be a bit more flexible. As this is basically just a function under the hood, makes sense to me.